OpenRA.Mods.Common/FileFormats/Png.cs

```diff
@@ -0,0 +1,213 @@
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Drawing.Imaging;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Net;
+using System.Text;
+
+namespace OpenRA.FileFormats
+{
+	public class Png
+	{
+		public int Width { get; set; }
+		public int Height { get; set; }
+		public PixelFormat PixelFormat { get; set; }
+		public Color[] Palette { get; set; }
+		public byte[] Data { get; set; }
+		public Dictionary<string, string> Meta = new Dictionary<string, string>();
+
+		public Png(Stream s)
+		{
+			if (!Verify(s))
+				throw new InvalidDataException("PNG Signature is bogus");
+
+			s.Position += 8;
+			var headerParsed = false;
+
+			using (var br = new BinaryReader(s))
+			{
+				var data = new List<byte>();
+
+				for (;;)
+				{
+					var length = IPAddress.NetworkToHostOrder(br.ReadInt32());
+					var type = Encoding.UTF8.GetString(br.ReadBytes(4));
+					var content = br.ReadBytes(length);
+					/*var crc = */br.ReadInt32();
+
+					if (!headerParsed && type != "IHDR")
+						throw new InvalidDataException("Invalid PNG file - header does not appear first.");
+
+					using (var ms = new MemoryStream(content))
+					{
+						switch (type)
+						{
+							case "IHDR":
+							{
+								if (headerParsed)
+									throw new InvalidDataException("Invalid PNG file - duplicate header.");
+								Width = IPAddress.NetworkToHostOrder(ms.ReadInt32());
+								Height = IPAddress.NetworkToHostOrder(ms.ReadInt32());
+								Data = new byte[Width * Height];
+
+								var bitDepth = ms.ReadUInt8();
+								var colorType = (PngColorType)ms.ReadByte();
+								PixelFormat = MakePixelFormat(bitDepth, colorType);
+
+								var compression = ms.ReadByte();
+								/*var filter = */ms.ReadByte();
+								var interlace = ms.ReadByte();
+
+								if (compression != 0)
+									throw new InvalidDataException("Compression method not supported");
+
+								if (interlace != 0)
+									throw new InvalidDataException("Interlacing not supported");
+
+								headerParsed = true;
+
+								break;
+							}
+
+							case "PLTE":
+							{
+								Palette = new Color[256];
+								for (var i = 0; i < length / 3; i++)
+								{
+									var r = ms.ReadByte(); var g = ms.ReadByte(); var b = ms.ReadByte();
+									Palette[i] = Color.FromArgb(r, g, b);
+								}
+
+								break;
+							}
+
+							case "tRNS":
+							{
+								if (Palette == null)
+									throw new InvalidDataException("Non-Palette indexed PNG are not supported.");
+
+								for (var i = 0; i < length; i++)
+									Palette[i] = Color.FromArgb(ms.ReadByte(), Palette[i]);
+
+								break;
+							}
+
+							case "IDAT":
+							{
+								data.AddRange(content);
+
+								break;
+							}
+
+							case "tEXt":
+							{
+								var key = new List<byte>();
+								byte b;
+
+								while ((b = ms.ReadUInt8()) != 0x00)
```

Comment:
This can simplify down to
```csharp
case "tEXt":
{
	var key = ms.ReadASCIIZ();
	Meta.Add(key, ms.ReadASCII(length - key.Length - 1));
	break;
}
```